lib/srv/authhandlers_test.go

```diff
@@ -223,3 +231,178 @@ func TestRBAC(t *testing.T) {
 		})
 	}
 }
+
+// TestRBACJoinMFA tests that MFA is enforced correctly when joining
+// sessions depending on the cluster auth preference and roles presented.
+func TestRBACJoinMFA(t *testing.T) {
+	t.Parallel()
+
+	const clusterName = "localhost"
+	const username = "testuser"
+
+	// create User CA
+	userTA := testauthority.New()
+	userCAPriv, err := userTA.GeneratePrivateKey()
+	require.NoError(t, err)
+	userCA, err := types.NewCertAuthority(types.CertAuthoritySpecV2{
+		Type:        types.UserCA,
+		ClusterName: clusterName,
+		ActiveKeys: types.CAKeySet{
+			SSH: []*types.SSHKeyPair{
+				{
+					PublicKey:      userCAPriv.MarshalSSHPublicKey(),
+					PrivateKey:     userCAPriv.PrivateKeyPEM(),
+					PrivateKeyType: types.PrivateKeyType_RAW,
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// create mock SSH server and add a cluster name
+	server := newMockServer(t)
+	cn, err := types.NewClusterName(types.ClusterNameSpecV2{
+		ClusterName: clusterName,
+		ClusterID:   "cluster_id",
+	})
+	require.NoError(t, err)
+	err = server.auth.SetClusterName(cn)
+	require.NoError(t, err)
+	ctx := context.Background()
+
+	accessPoint := &mockCAandAuthPrefGetter{
+		AccessPoint: server.auth,
+		cas: map[types.CertAuthType]types.CertAuthority{
+			types.UserCA: userCA,
+		},
+	}
+
+	// create auth handler and dummy node
+	config := &AuthHandlerConfig{
+		Server:      server,
+		Emitter:     &eventstest.MockEmitter{},
+		AccessPoint: accessPoint,
+	}
+	ah, err := NewAuthHandlers(config)
+	require.NoError(t, err)
+
+	node, err := types.NewServer("testie_node", types.KindNode, types.ServerSpecV2{
+		Addr:     "1.2.3.4:22",
+		Hostname: "testie",
+		Version:  types.V2,
+	})
+	require.NoError(t, err)
+
+	mfaAuthPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		SecondFactor:   constants.SecondFactorOTP,
+		RequireMFAType: types.RequireMFAType_SESSION,
```

Comment:
This was changed from `RequireMFAType_HARDWARE_KEY_TOUCH` because of logic here: https://github.com/gravitational/teleport/blob/dfa690acdad5f56ff19ff193e3cb876f907e98f7/lib/services/role.go#L1198

@codingllama why does that return `MFARequiredNever` in v13?

Reply:
I'm not sure, but I would guess it's some sort of backwards-compatible or transitional behavior. @Joerger might know.

Reply:
2 reasons:
- Hardware Key policies are enforced separately by the Authorizer, so any requirements should already be met by this point
- If the user passes the hardware_key_touch requirement, that means they are using a PIV key and already performed the touch at the start of the connection

Reply:
Gotcha, thanks for the explanation!